src/main/scala/mesosphere/marathon/api/LeaderProxyFilter.scala

```diff
@@ -56,66 +57,56 @@ class LeaderProxyFilter @Inject() (
       }
     }
 
-  @tailrec
+  private sealed trait LeaderResult
+  private case object WeAreLeader extends LeaderResult
+  private case object NoLeaderOrNotReady extends LeaderResult
+  private case class LeaderKnown(address: String) extends LeaderResult
+
+  @tailrec private def waitForConsistentLeadership(retries: Int = 10): LeaderResult = {
+    var result = false
+    val weAreLeader = electionService.isLeader
+    val leaderReady = marathonSchedulerService.leaderReady
+    lazy val currentLeaderData = electionService.leaderHostPort.filterNot(_ == myHostPort)
+
+    if (weAreLeader && leaderReady) {
+      WeAreLeader
+    } else if ((!weAreLeader) && (currentLeaderData.nonEmpty)) {
```

Comment:
How about refactoring this predicates into separate methods to make it easier to understand?